### PR TITLE
Fix Hoare optimizer failure for successive gates of three

### LINKED ISSUE #7271: Hoare optimizer fails for some circuits
### Environment

- **Qiskit Terra version**:  0.19.0.dev0+a603740
- **Python version**:  3.8.5
- **Operating system**: Linux

### What is happening?

Certain simple circuits fail raising a ValueError in `_multigate_opt` when trying to remove a node from its self.gatecache. 

I came across this when running random circuits through the pass and failures seem to occur when there are repeated gates at the beginning or end of the circuit.

### How can we reproduce the issue?

```python
from qiskit import QuantumCircuit
from qiskit.converters import circuit_to_dag
import qiskit.transpiler.passes.optimization.hoare_opt as hoare
qc = QuantumCircuit(1)
qc.h(0)
qc.h(0)
qc.h(0)
qc.rx(0.1, 0)
hoare.HoareOptimizer().run(circuit_to_dag(qc))
```

### What should happen?

Shouldn't get an exception.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by yjt98765**
### Summary
This PR fixes the bug reported by #7271.
When three H gates apply on the same qubit successively, the H gate in the middle would be deleted twice by the Hoare optimizer. It causes an exception.

### Details and comments
The function [`_multigate_opt`](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passes/optimization/hoare_opt.py#L324) examines the sequences of successive gates generated by `_target_successive_seq`. If a sequence combines to the identity, all the gates in the sequence are removed. It works for most cases except three successive H gates. See this example proposed by @ewinston:
```python
from qiskit import QuantumCircuit
from qiskit.converters import circuit_to_dag
import qiskit.transpiler.passes.optimization.hoare_opt as hoare
qc = QuantumCircuit(1)
qc.h(0)
qc.h(0)
qc.h(0)
hoare.HoareOptimizer().run(circuit_to_dag(qc))
```
I will use `H1`, `H2`, and `H3` to denote the `DAGOpNode` corresponding to the three H gates. The sequences generated by `_target_successive_seq` regarding `qubit[0]` is `[[H1, H2], [H2, H3]]`. After analyzing the first sequence (`[H1, H2]`), the optimizer deleted `H1` and `H2`. After analyzing the second sequence (`[H2, H3]`), the optimizer tried to delete these two gates and found that the key related to `H2` does not exist ().

To resolve this issue, I have moved the gate-deleting code from `_multigate_opt` to `_target_successive_seq`. When two gates are removed, the iteration index adds an extra 1, so that the deleted gate will not be considered in the next iteration. Since the function of `_target_successive_seq` has changed, I renamed it as `_remove_successive_identity`.

I have added a [test case](https://github.com/Qiskit/qiskit-terra/compare/main...yjt98765:hotfix7271?expand=1#diff-0a8f3d0ec10d806123fd7ef5c9f352731d294d199003c26d95de8b203187c97bR178-R197) to demonstrate the change. The test case is adapted based on @ewinston 's example and existing test cases. Before the change, the test case would fail with a `KeyError`. It passes after the modifications.

Close #7271

### CODE REVIEW

**Comment by ewinston** [comment]
What if instead you move the condition,
```python
if self._is_identity(seq) and self._seq_as_one(seq):
```
to `_target_successive_seq`
and then just check set intersection between new pair and last added pair? Something like,
```python
new_seq = [node1, node2]
if append:
    if self._is_identity(new_seq) and self._seq_as_one(new_seq):
        if not (seqs and set(new_seq) & set(seqs[-1])):
             seqs.append(new_seq)
```

**Comment by yjt98765** [comment]
> What if instead you move the condition,
> 
> ```python
> if self._is_identity(seq) and self._seq_as_one(seq):
> ```
> 
> to `_target_successive_seq` and then just check set intersection between new pair and last added pair? Something like,
> 
> ```python
> new_seq = [node1, node2]
> if append:
>     if self._is_identity(new_seq) and self._seq_as_one(new_seq):
>         if not (seqs and set(new_seq) & set(seqs[-1])):
>              seqs.append(new_seq)
> ```
> 
> This way you only have to check the last added pair.

Yes, it seems better. Maybe the whole block can be moved to `_target_successive_seq`:
```
                for node in seq:
                    dag.remove_op_node(node)
                    removed.add(node)
                    # if recursive call, gate will be removed further down
                    if max_idx is None:
                        for qbt in node.qargs:
                            self.gatecache[qbt].remove(node)
                    else:
                        if self.gatecache[qubit].index(node) > max_idx:
                            for qbt in node.qargs:
                                self.gatecache[qbt].remove(node)
```
It is not related to the other parts of `_multigate_opt`, but close to `_target_successive_seq`. After the moving, `_target_successive_seq` can be renamed as something like `_remove_successive_identity`.

**Comment by ewinston** [comment]
@yjt98765 That sounds good.

**Comment by ewinston** [comment]
@yjt98765 Thanks for the fix!

**Review by ewinston — DISMISSED** [review]

File: qiskit/transpiler/passes/optimization/hoare_opt.py [hunk 1]
```diff
@@ -232,23 +232,24 @@ def _traverse_dag(self, dag):
...
+            other immediately on these target qubits, and combine to the
+            identity (consider sequences of length 2 for now)
         Args:
+            dag (DAGCircuit): the directed acyclic graph to run on.
             qubit (Qubit): qubit cache to inspect
-        Returns:
-            list(list(DAGOpNode or DAGOutNode)): list of target successive gate sequences for
-                                 this qubit's cache
+            max_idx (int): a value indicates a recursive call, optimize
+                           and remove gates up to this point in the cache
```
Comment (jakelishman) [inline qiskit/transpiler/passes/optimization/hoare_opt.py:243]: This documentation is a bit confusing now - `_target_successive_seq` doesn't call itself recursively.

File: qiskit/transpiler/passes/optimization/hoare_opt.py [hunk 2]
```diff
@@ -260,10 +261,19 @@ def _target_successive_seq(self, qubit):
...
+                for node in seq:
+                    dag.remove_op_node(node)
+                    # if recursive call, gate will be removed further down
+                    if max_idx is None:
+                        for qbt in node.qargs:
+                            self.gatecache[qbt].remove(node)
+                    else:
+                        if self.gatecache[qubit].index(node) > max_idx:
+                            for qbt in node.qargs:
+                                self.gatecache[qbt].remove(node)
```
Comment (jakelishman) [inline qiskit/transpiler/passes/optimization/hoare_opt.py:276]: A similar comment here - `_remove_successive_identity` isn't a recursive function.

Also, I know this was pre-existing, but this whole logic block is equivalent to
```python
if max_idx is None or self.gatecache[qubit].index(node) > max_idx:
    for qbt in node.qargs:
        self.gatecache[qbt].remove(node)
```
which avoids an extra scope, and removes some duplication.

(A comment far beyond the scope of the PR: this pre-existing code smells of poor scaling - `list.index` and `list.remove` are very tricky to use well, and this is highly likely to be at least quadratic in the circuit depth.)

**Review by jakelishman — CHANGES_REQUESTED** [review]
Thanks for the fix, and sorry for the delay!  I had a couple more comments down below - don't worry about the bit that says it's out-of-scope, that's just for history purposes.  Please could you also add a release note describing the bugfix as well?

File: qiskit/transpiler/passes/optimization/hoare_opt.py (hunk 1 above)
Comment (yjt98765) [reply qiskit/transpiler/passes/optimization/hoare_opt.py]: I see your point here: the moved code/comment should make sense in the new context. I have updated the docstring. In addition, `max_idx` is a bit misleading here, so I renamed it to `from_idx`.

File: qiskit/transpiler/passes/optimization/hoare_opt.py (hunk 2 above)
Comment (yjt98765) [reply qiskit/transpiler/passes/optimization/hoare_opt.py]: Indeed, I have merged the code as you suggested here.

**Comment by yjt98765** [comment]
@jakelishman sure, I have added a release note.

**Comment by jakelishman** [comment]
I pushed a commit just to reword the release note a little - mostly I wanted to avoid referring to a private function in it, because users might not know what that meant.

I was about to merge this, but then I realised I'm a bit confused - the `max_idx` is now called `from_idx`, which makes sense to me for how it's used, but the whole function seems to use it as the minimum index (and it's still called `max_idx` in the parent functions).  Is this intended?  I'm not really sure what's going on here.  If it all _is_ meant to be the minimum index not the maximum, then it'd be good to change the names everywhere to match.

**Comment by yjt98765** [comment]
@jakelishman I am not the author of the original program, but I guess the key to understanding the parameter is the last line of the parent function `_multigate_opt`:
```
self.gatecache[qubit] = self.gatecache[qubit][max_idx + 1 :]
```
The variable `max_idx` is a boundary between the removed and the remained indexes. It is the *maximum* index to be removed. Its next index is the *minimum* index to be kept. So `max_idx` and `min_idx` both make some sense and both cause some confusion. Maybe we can call it `max_tbr_idx` or something more accurate?

**Review by jakelishman — APPROVED** [review]
Thanks - it seems ok to me to leave it as it is, then.

**Pull request merged**